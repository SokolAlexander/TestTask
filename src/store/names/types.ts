export enum REQUEST_STATUS {
  'IDLE',
  'PENDING',
  'SUCCESS',
  'FAILURE',
}

export type Gender = 'male' | 'female' | '';

export type NamesState = {
  request: {
    status: REQUEST_STATUS;
    error: string | null;
  };
  name: string;
  gender: Gender;
  stats: {
    countFemale: number;
    history: {
      [name: string]: number;
    };
  };
};

export type GenderData = {
  name: string;
  gender: Gender;
  probability: number;
  count: number;
};

export interface BaseAction {
  type: string;
}

export interface GetGenderSuccess extends BaseAction {
  payload: GenderData;
}
export interface GetGenderFailure extends BaseAction {
  payload: string;
}

export type NamesActions = BaseAction | GetGenderFailure | GetGenderSuccess;
