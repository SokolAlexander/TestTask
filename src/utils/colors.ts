export const colors = {
  black: '#000000',
  white: '#FFFFFF',
  darkGrey: '#141513',
  grey: '#DDD',
  lightGrey: '#EDEDED',
};
