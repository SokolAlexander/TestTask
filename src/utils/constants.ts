export const API_URL = 'https://api.genderize.io?name=';
