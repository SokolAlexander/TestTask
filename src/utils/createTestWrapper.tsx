import React from 'react';
import thunk from 'redux-thunk';
import createMockStore from 'redux-mock-store';

import {NamesState, REQUEST_STATUS} from '../store/names/types';
import {Provider} from 'react-redux';

export const defaultMockState: NamesState = {
  name: '',
  gender: 'female',
  request: {
    status: REQUEST_STATUS.IDLE,
    error: null,
  },
  stats: {
    history: {},
    countFemale: 0,
  },
};

export const makeStore = createMockStore([thunk]);
export const defaultMockStore = makeStore(defaultMockState);

export const createTestWrapper = (mockStore = defaultMockStore) => {
  return ({children}: {children: React.ReactElement}) => (
    <Provider store={mockStore}>{children}</Provider>
  );
};
