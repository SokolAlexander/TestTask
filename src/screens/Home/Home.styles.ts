import {StyleSheet} from 'react-native';
import {colors} from '../../utils/colors';

export const homeStyles = StyleSheet.create({
  wrapper: {
    flex: 1,
    padding: 20,
    paddingTop: 40,
    backgroundColor: colors.white,
  },
  button: {
    position: 'absolute',
    bottom: 0,
    left: 0,
    right: 0,
    backgroundColor: colors.darkGrey,
    flex: 1,
    alignItems: 'center',
    justifyContent: 'center',
    padding: 25,
  },
  btnText: {
    color: colors.white,
    fontSize: 22,
    fontWeight: 'bold',
  },
  safeArea: {
    flex: 1,
    backgroundColor: colors.darkGrey,
  },
});
