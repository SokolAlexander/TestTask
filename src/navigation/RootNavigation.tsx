import React from 'react';
import {createNativeStackNavigator} from '@react-navigation/native-stack';

import {RootStackParamList} from './RootNavigation.types';
import {Home} from '../screens/Home/Home';
import {Statistics} from '../screens/Statistics/Statistics';

const RootStack = createNativeStackNavigator<RootStackParamList>();

export const RootNavigation = () => {
  return (
    <RootStack.Navigator>
      <RootStack.Screen name="Home" component={Home} />
      <RootStack.Screen name="Statistics" component={Statistics} />
    </RootStack.Navigator>
  );
};
