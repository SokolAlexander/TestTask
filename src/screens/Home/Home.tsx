import React from 'react';
import {Text, TouchableOpacity, View} from 'react-native';
import {SafeAreaView} from 'react-native-safe-area-context';

import {Form} from '../../components/Form/Form';
import {homeStyles} from './Home.styles';
import {HomeProps} from './Home.types';

export const Home = ({navigation}: HomeProps) => {
  const goToStats = () => {
    navigation.push('Statistics');
  };

  return (
    <SafeAreaView style={homeStyles.safeArea} edges={['bottom']}>
      <View style={homeStyles.wrapper}>
        <Form />
        <TouchableOpacity
          accessibilityRole="button"
          style={homeStyles.button}
          onPress={goToStats}>
          <Text style={homeStyles.btnText}>To Statistics</Text>
        </TouchableOpacity>
      </View>
    </SafeAreaView>
  );
};
