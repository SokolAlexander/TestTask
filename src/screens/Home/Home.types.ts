import {NativeStackNavigationProp} from '@react-navigation/native-stack';
import {RootStackParamList} from '../../navigation/RootNavigation.types';

export type HomeProps = {
  navigation: NativeStackNavigationProp<RootStackParamList, 'Home'>;
};
