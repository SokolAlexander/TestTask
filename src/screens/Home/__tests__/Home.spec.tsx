import {NativeStackNavigationProp} from '@react-navigation/native-stack';
import {fireEvent, render} from '@testing-library/react-native';
import React from 'react';
import {RootStackParamList} from '../../../navigation/RootNavigation.types';
import {createTestWrapper} from '../../../utils/createTestWrapper';
import {Home} from '../Home';

const DefaultTestWrapper = createTestWrapper();
const mockNavigation = {
  push: jest.fn(),
} as any as NativeStackNavigationProp<RootStackParamList, 'Home'>;

describe('Home screen', () => {
  it('matches snapshot', () => {
    const component = render(
      <DefaultTestWrapper>
        <Home navigation={mockNavigation} />
      </DefaultTestWrapper>,
    );

    expect(component).toMatchSnapshot();
  });

  it('navigates to Statistics screen when btn pressed', () => {
    const {getByText} = render(
      <DefaultTestWrapper>
        <Home navigation={mockNavigation} />
      </DefaultTestWrapper>,
    );

    const btn = getByText('To Statistics');
    fireEvent(btn, 'press');
    expect(mockNavigation.push).toHaveBeenCalledTimes(1);
    expect(mockNavigation.push).toHaveBeenCalledWith('Statistics');
  });
});
