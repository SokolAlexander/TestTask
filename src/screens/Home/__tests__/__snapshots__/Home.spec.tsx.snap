// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home screen matches snapshot 1`] = `
<RNCSafeAreaView
  edges={
    Array [
      "bottom",
    ]
  }
  style={
    Object {
      "backgroundColor": "#141513",
      "flex": 1,
    }
  }
>
  <View
    style={
      Object {
        "backgroundColor": "#FFFFFF",
        "flex": 1,
        "padding": 20,
        "paddingTop": 40,
      }
    }
  >
    <TextInput
      onChangeText={[Function]}
      placeholder="Enter a name"
      style={
        Object {
          "backgroundColor": "#FFFFFF",
          "borderColor": "#141513",
          "borderWidth": 1,
          "marginVertical": 10,
          "paddingHorizontal": 15,
          "paddingVertical": 10,
        }
      }
      value=""
    />
    <View
      style={
        Object {
          "alignItems": "center",
          "minHeight": 30,
        }
      }
    />
    <View
      accessibilityRole="button"
      accessibilityState={
        Object {
          "disabled": true,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "#EDEDED",
          "borderColor": "#DDD",
          "borderWidth": 2,
          "justifyContent": "center",
          "marginHorizontal": 25,
          "marginTop": 10,
          "opacity": 1,
          "padding": 15,
        }
      }
    >
      <Text
        style={
          Object {
            "color": "#141513",
            "fontSize": 18,
          }
        }
      >
        Get Gender
      </Text>
    </View>
    <View
      accessibilityRole="button"
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "#141513",
          "bottom": 0,
          "flex": 1,
          "justifyContent": "center",
          "left": 0,
          "opacity": 1,
          "padding": 25,
          "position": "absolute",
          "right": 0,
        }
      }
    >
      <Text
        style={
          Object {
            "color": "#FFFFFF",
            "fontSize": 22,
            "fontWeight": "bold",
          }
        }
      >
        To Statistics
      </Text>
    </View>
  </View>
</RNCSafeAreaView>
`;
